tokio/src/net/windows/named_pipe.rs

```diff
@@ -0,0 +1,1280 @@
+    /// same or higher number of bytes available ([total_bytes_available]). It
+    /// might even report a count of `0` even if no data has been read from the
+    /// named pipe that was previously peeked.
+    ///
+    /// Peeking does not update the state of the named pipe, so in order to
+    /// advance it you have to actively issue reads. A peek reporting a number
+    /// of bytes available ([total_bytes_available]) of `0` does not guarantee
+    /// that there is no data available to read from the named pipe. Even if a
+    /// peer is writing data, reads still have to be issued for the state of the
+    /// named pipe to update.
+    ///
+    /// Finally, peeking might report no data available indefinitely if there's
+    /// too little data in the buffer of the named pipe.
+    ///
+    /// You can play around with the [`named-pipe-peek` example] to get a feel
+    /// for how this function behaves.
+    ///
+    /// [total_bytes_available]: PipePeekInfo::total_bytes_available
+    /// [`named-pipe-peek` example]: https://github.com/tokio-rs/tokio/blob/master/examples/named-pipe-peek.rs
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::io;
+    /// use tokio::io::{AsyncReadExt as _, AsyncWriteExt as _};
+    /// use tokio::net::windows::{NamedPipeOptions, NamedPipeClientOptions};
+    ///
+    /// const PIPE_NAME: &str = r"\\.\pipe\tokio-named-pipe-peek-consumed";
+    /// const N: usize = 100;
+    ///
+    /// # #[tokio::main] async fn main() -> std::io::Result<()> {
+    /// let mut server = NamedPipeOptions::new().create(PIPE_NAME)?;
+    /// let mut client = NamedPipeClientOptions::new().create(PIPE_NAME)?;
+    /// server.connect().await?;
+    ///
+    /// let client = tokio::spawn(async move {
+    ///     for _ in 0..N {
+    ///         client.write_all(b"ping").await?;
+    ///     }
+    ///
+    ///     let mut buf = [0u8; 4];
+    ///     client.read_exact(&mut buf).await?;
+    ///
+    ///     Ok::<_, io::Error>(buf)
+    /// });
+    ///
+    /// let mut buf = [0u8; 4];
+    /// let mut available = 0;
+    ///
+    /// for n in 0..N {
+    ///     if available < 4 {
+    ///         server.read_exact(&mut buf).await?;
+    ///         assert_eq!(&buf[..], b"ping");
+    ///
+    ///         let (_, info) = server.peek(None)?;
+    ///         available = info.total_bytes_available;
+    ///         continue;
+    ///     }
+    ///
+    ///     // here we know that at least `available` bytes are immediately
+    ///     // ready to read.
+    ///     server.read_exact(&mut buf).await?;
+    ///     available -= buf.len();
+    ///     assert_eq!(&buf[..], b"ping");
+    /// }
+    ///
+    /// server.write_all(b"pong").await?;
+    ///
+    /// let buf = client.await??;
+    /// assert_eq!(&buf[..], b"pong");
+    /// # Ok(()) }
+    /// ```
+    pub fn peek(&mut self, buf: Option<&mut [u8]>) -> io::Result<(usize, PipePeekInfo)> {
```

Comment:
Also, since we are defining a custom type for the return value, could the `usize` be included on that type?
```suggestion
    pub fn peek(&mut self, buf: Option<&mut [u8]>) -> io::Result<(usize, PipePeekInfo)> {
```

Reply:
*if* this function switches to `ReadBuf`, it won' need to return the `usize` anyway as that is handled by `ReadBuf`.

Reply:
Yeah. I also realized on this morning walk that it should be `ReadBuf`. Cheers!